src/inference/AdvancedSMC.jl

```diff
@@ -147,72 +224,44 @@ function step(model, spl::Sampler{<:PG}, vi::VarInfo)
-
-    return particles[indx].vi, true
-end
-
-function sample(  model::Model,
-                  alg::PG;
-                  save_state=false,         # flag for state saving
-                  resume_from=nothing,      # chain to continue
-                  reuse_spl_n=0             # flag for spl re-using
-                )
-
-    spl = reuse_spl_n > 0 ?
-          resume_from.info[:spl] :
-          Sampler(alg)
-    if resume_from != nothing
-        spl.selector = resume_from.info[:spl].selector
-    end
-    @assert typeof(spl.alg) == typeof(alg) "[Turing] alg type mismatch; please use resume() to re-use spl"
-
-    n = reuse_spl_n > 0 ?
-        reuse_spl_n :
-        alg.n_iters
-    samples = Vector{Sample}()
-
-    ## custom resampling function for pgibbs
-    ## re-inserts reteined particle after each resampling step
-    time_total = zero(Float64)
 
-    vi = resume_from == nothing ?
-        empty!(VarInfo(model)) :
-        resume_from.info[:vi]
+    # Extract the VarInfo from the retained particle.
+    params = tonamedtuple(spl.state.vi)
+    spl.state.vi = particles[indx].vi
+    lp = getlogp(spl.state.vi)
 
-    pm = nothing
-    PROGRESS[] && (spl.info[:progress] = ProgressMeter.Progress(n, 1, "[PG] Sampling...", 0))
-
-    for i = 1:n
-        time_elapsed = @elapsed vi, _ = step(model, spl, vi)
-        push!(samples, Sample(vi))
-        samples[i].value[:elapsed] = time_elapsed
-
-        time_total += time_elapsed
+    # update the master vi.
+    return transition(params, lp, particles.logE, 1.0)
+end
 
-        if PROGRESS[] && spl.selector.tag == :default
-            ProgressMeter.next!(spl.info[:progress])
-        end
-    end
+function sample_end!(
+    ::AbstractRNG,
+    ::Model,
+    spl::Sampler{<:ParticleInference},
+    N::Integer,
+    ts::Vector{ParticleTransition};
+    kwargs...
+)
+    # Set the default for resuming the sampler.
+    resume_from = get(kwargs, :resume_from, nothing)
 
-    @info("[PG] Finished with")
-    @info("  Running time    = $time_total;")
+    # Exponentiate the average log evidence.
+    loge = exp(mean([t.le for t in ts]))
 
-    loge = exp.(mean(spl.info[:logevidence]))
+    # If we already had a chain, grab the logevidence.
     if resume_from != nothing   # concat samples
-        pushfirst!(samples, resume_from.info[:samples]...)
+        @assert resume_from isa Chains "resume_from needs to be a Chains object."
+        # pushfirst!(samples, resume_from.info[:samples]...)
         pre_loge = exp.(resume_from.logevidence)
         # Calculate new log-evidence
-        pre_n = length(resume_from.info[:samples])
-        loge = (log(pre_loge) * pre_n + log(loge) * n) / (pre_n + n)
-    end
-    c = Chain(loge, samples)       # wrap the result by Chain
-
-    if save_state               # save state
-        c = save(c, spl, model, vi, samples)
+        pre_n = length(resume_from)
+        loge = (log(pre_loge) * pre_n + log(loge) * N) / (pre_n + N)
```

Comment:
Why do we need to do `pre_loge = exp.(resume_from.logevidence)` if we do log here?

Reply:
If `pre_loge = resume_from.logevidence`, consider changing to:

```julia
pre_loge = resume_from.logevidence
avgevidence = (pre_loge*pre_n + log(avgevidence)*N) / (pre_n + N)
```
or if we don't do a `exp` higher up...
```julia
pre_loge = resume_from.logevidence
loge = (pre_loge*pre_n + loge*N) / (pre_n + N)
```

Reply:
Yeah, that's way better.